Sources/NIOCore/EventLoopFuture.swift

```diff
@@ -1334,6 +1388,67 @@ extension EventLoopFuture {
             }
         }
     }
+
+    /// Loops through the futures array and attaches callbacks to execute `onValue` on the provided `EventLoop` when
+    /// they succeed. The `onValue` will receive the index of the future that fulfilled the provided `Result`.
+    ///
+    /// Once all the futures have succeed, the provided promise will succeed.
+    /// Once any future fails, the provided promise will fail.
+    @inlinable
+    internal static func _reduceSuccesses0(
+        _ promise: EventLoopPromise<Void>,
+        _ futures: [EventLoopFuture<Value>],
+        _ eventLoop: EventLoop
+    ) {
+        eventLoop.assertInEventLoop()
+
+        if futures.count == 0 {
+            promise.succeed(())
+            return
+        }
+
+        let remainingCount = NIOLoopBoundBox(_value: futures.count, uncheckedEventLoop: eventLoop)
+
+        // Sends the result to `onValue` in case of success and succeeds/fails the input promise, if appropriate.
+        @Sendable
+        func processResult(_ index: Int, _ result: Result<Void, Error>) {
+            switch result {
+            case .success:
+                remainingCount.value -= 1
+
+                if remainingCount.value == 0 {
+                    promise.succeed(())
+                }
+            case .failure(let error):
+                promise.fail(error)
+            }
+        }
+        // loop through the futures to chain callbacks to execute on the initiating event loop and grab their index
+        // in the "futures" to pass their result to the caller
+        for (index, future) in futures.enumerated() {
+            if future.eventLoop.inEventLoop,
+                let result = future._value {
+                // Fast-track already-fulfilled results without the overhead of calling `whenComplete`. This can yield a
+                // ~20% performance improvement in the case of large arrays where all elements are already fulfilled.
+                switch result {
+                case .success:
+                    processResult(index, .success(()))
+                case .failure(let error):
+                    processResult(index, .failure(error))
+                }
+                if case .failure = result {
+                    return  // Once the promise is failed, future results do not need to be processed.
+                }
+            } else {
+                // We have to map to `Void` here to avoid sharing the potentially non-Sendable
+                // value across event loops.
+                future
+                    .map { _ in () }
+                    .hop(to: eventLoop)
+                    .whenComplete { result in processResult(index, result) }
```

Comment:
Can't we do something like the following to avoid the extra ELFs?

```swift
future.whenComplete { result in
  let voidResult = result.map { _ in }
  future.eventLoop.exeute {
    processResult(index, voidResult)
  }
}
```